File: backend/app/models/objectifs.py
import datetime as dt

from sqlmodel import Field, SQLModel


class LongTermGoal(SQLModel, table=True):
    """Objectif long terme : masters, concours, opportunités carrière."""

    __tablename__ = "long_term_goal"
    id: int | None = Field(default=None, primary_key=True)
    titre: str
    categorie: str = "autre"  # master|concours|carriere|autre
    statut: str = "veille"    # veille|preparation|candidature|obtenu|abandonne
    echeance: dt.date | None = None
    progression: int = 0      # 0-100
    description: str | None = None
    lien: str | None = None
    cree_le: dt.date = Field(default_factory=dt.date.today)

File: frontend/lib/objectifs.ts
// Types + client API pour le module Objectifs long terme (proxy Next -> backend)
import { json } from "./fetch-json";

const BASE = "/api/objectifs";

export interface LongTermGoal {
  id: number;
  titre: string;
  categorie: "master" | "concours" | "carriere" | "autre";
  statut: "veille" | "preparation" | "candidature" | "obtenu" | "abandonne";
  echeance?: string | null;
  progression: number;
  description?: string | null;
  lien?: string | null;
  cree_le: string;
}

export const objectifsApi = {
  list: (): Promise<LongTermGoal[]> => fetch(`${BASE}/goals`).then((r) => json<LongTermGoal[]>(r)),
  create: (data: Partial<LongTermGoal>): Promise<LongTermGoal> =>
    fetch(`${BASE}/goals`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }).then((r) => json<LongTermGoal>(r)),
  update: (id: number, data: Partial<LongTermGoal>): Promise<LongTermGoal> =>
    fetch(`${BASE}/goals/${id}`, {
      method: "PATCH",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(data),
    }).then((r) => json<LongTermGoal>(r)),
  remove: (id: number): Promise<Response> => fetch(`${BASE}/goals/${id}`, { method: "DELETE" }),
};
